### ISSUE TITLE
[Escalating Issues] Generic Metrics Backend dataset not equal to Errors dataset

### ISSUE THREAD

**Issue by NisanthanNanthakumar** [getsentry/snuba#4923]
### Environment

Production Sentry SaaS

### Expected Result

I expect the query to the Generic Metrics Backend and the query to the Errors dataset to return the same hourly event counts.

### Actual Result

We have differing values. The [GCP logs](https://console.cloud.google.com/logs/query;query=jsonPayload.event%3D%22Generics%20Metrics%20Backend%20query%20results%20not%20the%20same%20as%20Errors%20dataset%20query.%22;cursorTimestamp=2023-10-25T00:05:46.529913154Z;duration=PT1H?project=internal-sentry&pli=1&rapt=AEjHL4OaLO_Bc9sxNHRLpxt0cv_48CLDJ8W6od8kQuqWCoDwXaw46UlXzAdmLyONfwTY1qH1HMcG67P2TIFxQA0F-gKnH7JfGA&authuser=1) show that the results from Metrics Backend query and Errors Dataset query are not equal.

We confirmed that the queries for both Metrics and Errors are correct using prod Snuba Admin.

One potential reason for the discrepancy is that the timestamp used in the Generic Metrics Backend is from `post_process`, whereas Errors dataset uses the timestamp from the sdk. 

However, I'd expect only small differences in hourly bucket values not the drastically different counts that I am currently getting. 

Code where we make the queries: https://github.com/getsentry/sentry/blob/e52836d69b21481ab9d27e271666c1d887aad04e/src/sentry/issues/escalating.py#L159

**Comment by evanh** [comment]
@NisanthanNanthakumar 

Thank you! I found some log entries, but they all seem explainable by small timestamp discrepancies. For example:

```
Time: Metrics/Dataset
Time: Metrics/Dataset
2023-11-02T00:00:00: 10/10
2023-11-02T04:00:00: 1/1
2023-11-02T05:00:00: 2/2
2023-11-02T06:00:00: 1/1
2023-11-02T08:00:00: 10/10
2023-11-02T09:00:00: 5/5
2023-11-02T11:00:00: 34/34
2023-11-02T12:00:00: 2/2
2023-11-02T14:00:00: 36/38
2023-11-02T15:00:00: 119/117
2023-11-02T16:00:00: 206/206
2023-11-02T17:00:00: 30/30
2023-11-02T18:00:00: 49/52
```

Here `2023-11-02T14:00:00` and `2023-11-02T15:00:00` are both off, but they're both off by 2 and in the opposite direction (-2/+2). Are there are any examples you can point me too that are different from this?

**Comment by evanh** [comment]
Something else I wanted to bring up: I think escalating issues would be a good candidate for using the new metrics layer. The metrics query you're doing seems very simple and could easily be migrated.

The entrypoint for the new metrics layer is the `run_query` function found [here](https://github.com/getsentry/sentry/blob/master/src/sentry/snuba/metrics_layer/query.py#L38). You can see some examples of how it works [here](https://github.com/getsentry/sentry/blob/master/tests/snuba/test_metrics_layer.py#L75). If this is something you're willing to adopt please let me know, I'm happy to help and get any feedback you might have.

The idea of the new layer is to make using metrics easier, so hopefully you could get some use out of it.

**Comment by evanh** [comment]
> From what we discussed, it sounds like Errors dataset uses the timestamp provided by the SDKs and the Generic Metrics Backend uses the timestamp of when we process the event in post_process.

Well then yes, the results won't line up exactly. Is this something you need for your feature? I am not seeing anything unexpected here.

> But I am assuming that this runs on top of the Generic Metrics Backend?

Yes it's the same backend but a much simpler way of writing queries (no result modifiers, derived metrics etc.)

**Comment by ayirr7** [comment]
```
     errors_count  metrics_count  (errors - metrics)
0              16           16.0                 0.0
1              12           12.0                 0.0
2              30           30.0                 0.0
3              33           33.0                 0.0
4              39           39.0                 0.0
5             103          103.0                 0.0
6              38           38.0                 0.0
7              21           21.0                 0.0
8               9            9.0                 0.0
9               2            2.0                 0.0
10              1            1.0                 0.0
11              4            4.0                 0.0
12              1            1.0                 0.0
13              3            3.0                 0.0
15              1            1.0                 0.0
16              2            2.0                 0.0
17             10           10.0                 0.0
18              3            3.0                 0.0
19              3            3.0                 0.0
20              9            8.0                 1.0
21             10           10.0                 0.0
22              1            1.0                 0.0
23              2            2.0                 0.0
24              2            2.0                 0.0
25              1            1.0                 0.0
26              7            7.0                 0.0
27              1            1.0                 0.0
28              4            4.0                 0.0
29              4            4.0                 0.0
30             11           11.0                 0.0
31             12           12.0                 0.0
32             12           11.0                 1.0
33              1            2.0                -1.0
34              1            1.0                 0.0
35              5            5.0                 0.0
36              2            2.0                 0.0
37             16           16.0                 0.0
38              6            6.0                 0.0
39              7            7.0                 0.0
40              2            2.0                 0.0
41              1            1.0                 0.0
42              5            5.0                 0.0
43             30           30.0                 0.0
44              4            4.0                 0.0
45              5            5.0                 0.0
46              4            4.0                 0.0
47              1            1.0                 0.0
48             34           34.0                 0.0
49              4            4.0                 0.0
50              1            1.0                 0.0
51             24           24.0                 0.0
52             14           14.0                 0.0
53              1            1.0                 0.0
54              3            3.0                 0.0
55              3            3.0                 0.0
56              1            1.0                 0.0
57              1            1.0                 0.0
58              2            2.0                 0.0
59             10           10.0                 0.0
```

@NisanthanNanthakumar I ran some data analysis to check the differences between some of the errors and metrics counts, and a small sample is pasted above. I have the full table in a separate document if you're interested in seeing that. 

From the sample above and the full table data, it appears that the discrepancies are small and very possibly related to the timestamp discrepancy we had discussed. 

PR for the fix is here and will be merged on Monday. https://github.com/getsentry/sentry/pull/57435

**Comment by evanh** [comment]
@NisanthanNanthakumar We'd like to understand better if lining up the data between the two datasets is a requirement of escalating issues or a debugging concern. We're trying to find the best way forward to help escalating issues. Can you clarify: Is it a requirement that the two datasets line up exactly?

**Comment by ayirr7** [comment]
@NisanthanNanthakumar Here's the sorted hourly bucket list for the log event you have linked. As you can see, the hourly buckets are not consecutive, in fact, many hours are skipped on each day. I think the logs may need to be built and emitted in a different way, because the logs currently do not show consecutive hourly buckets. 

```
                   hourBucket  
2023-10-31T14:00:00+00:00          
2023-10-31T16:00:00+00:00           
2023-10-31T17:00:00+00:00           
2023-10-31T19:00:00+00:00           
2023-11-01T00:00:00+00:00           
2023-11-01T02:00:00+00:00         
2023-11-01T05:00:00+00:00           
2023-11-01T09:00:00+00:00           
2023-11-01T09:00:00+00:00          
2023-11-01T11:00:00+00:00           
2023-11-01T15:00:00+00:00           
2023-11-01T16:00:00+00:00           
2023-11-01T17:00:00+00:00         
2023-11-01T18:00:00+00:00           
2023-11-01T19:00:00+00:00           
2023-11-01T21:00:00+00:00          
2023-11-02T04:00:00+00:00           
2023-11-02T05:00:00+00:00           
2023-11-02T07:00:00+00:00           
2023-11-02T10:00:00+00:00           
2023-11-02T11:00:00+00:00           
2023-11-02T14:00:00+00:00           
2023-11-02T15:00:00+00:00           
2023-11-02T15:00:00+00:00          
2023-11-02T17:00:00+00:00          
2023-11-02T18:00:00+00:00          
2023-11-02T20:00:00+00:00           
 2023-11-02T21:00:00+00:00           
 2023-11-02T23:00:00+00:00           
2023-11-03T01:00:00+00:00           
2023-11-03T05:00:00+00:00           
2023-11-03T07:00:00+00:00           
2023-11-03T08:00:00+00:00           
2023-11-03T08:00:00+00:00           
2023-11-03T09:00:00+00:00           
2023-11-03T12:00:00+00:00            
2023-11-03T13:00:00+00:00           
2023-11-03T14:00:00+00:00           
2023-11-03T15:00:00+00:00           
2023-11-03T16:00:00+00:00           
2023-11-03T17:00:00+00:00           
2023-11-03T19:00:00+00:00           
2023-11-04T02:00:00+00:00           
2023-11-04T09:00:00+00:00          
2023-11-04T09:00:00+00:00           
2023-11-04T10:00:00+00:00          
2023-11-04T14:00:00+00:00           
2023-11-04T15:00:00+00:00           
2023-11-04T19:00:00+00:00           
2023-11-04T21:00:00+00:00           
 2023-11-04T23:00:00+00:00           
2023-11-05T00:00:00+00:00           
2023-11-05T08:00:00+00:00           
2023-11-05T11:00:00+00:00           
2023-11-05T13:00:00+00:00           
2023-11-05T13:00:00+00:00           
2023-11-05T14:00:00+00:00           
2023-11-05T23:00:00+00:00           
2023-11-06T04:00:00+00:00           
 2023-11-06T06:00:00+00:00           
2023-11-06T08:00:00+00:00           
2023-11-06T11:00:00+00:00           
2023-11-06T12:00:00+00:00           
2023-11-06T14:00:00+00:00           
 2023-11-07T07:00:00+00:00           
2023-11-07T09:00:00+00:00           
2023-11-07T11:00:00+00:00           
```

**Comment by evanh** [comment]
> I think it is ok if the difference between the datasets for an hourly bucket is reconciled in the next hourly bucket.

So does that mean the issue is solved? As we mentioned in the ticket earlier, the datasets are using different timestamps so the numbers are basically guaranteed to not line up.

Can you clarify the requirement that the data match exactly?

**Comment by ayirr7** [comment]
@NisanthanNanthakumar The reason why we are seeing 35 vs 46 buckets is because, if a bucket has a count of 0, the logs are currently not emitting a metric. So what is happening, is that the buckets that have a count of 0 are omitted. 

Here's the full result for [this log event](https://console.cloud.google.com/logs/query;query=jsonPayload.event%3D%22Generics%20Metrics%20Backend%20query%20results%20not%20the%20same%20as%20Errors%20dataset%20query.%22%0Atimestamp%3D%222023-11-07T12:09:22.687026181Z%22%0AinsertId%3D%22t4ovb5y83lsq9fvm%22;cursorTimestamp=2023-11-07T12:09:22.687026181Z;startTime=2023-11-07T09:37:08.571Z;endTime=2023-11-07T17:38:08.571Z?project=internal-sentry): 
```

                   hourBucket  errors_count  metrics_count  (errors - metrics)
2023-10-31T14:00:00+00:00           1.0            NaN                 NaN
2023-10-31T16:00:00+00:00           NaN            1.0                 NaN
2023-10-31T17:00:00+00:00           1.0            1.0                 0.0
2023-10-31T19:00:00+00:00           1.0            NaN                 NaN
2023-11-01T00:00:00+00:00           1.0            NaN                 NaN
2023-11-01T02:00:00+00:00           NaN            1.0                 NaN
2023-11-01T05:00:00+00:00           NaN            1.0                 NaN
2023-11-01T09:00:00+00:00           NaN            2.0                 NaN
2023-11-01T09:00:00+00:00           NaN            1.0                 NaN
2023-11-01T11:00:00+00:00           1.0            NaN                 NaN
2023-11-01T15:00:00+00:00           NaN            1.0                 NaN
2023-11-01T16:00:00+00:00           NaN            1.0                 NaN
2023-11-01T17:00:00+00:00           1.0            NaN                 NaN
2023-11-01T18:00:00+00:00           1.0            1.0                 0.0
2023-11-01T19:00:00+00:00           1.0            NaN                 NaN
2023-11-01T21:00:00+00:00           1.0            NaN                 NaN
2023-11-02T04:00:00+00:00           1.0            NaN                 NaN
2023-11-02T05:00:00+00:00           NaN            1.0                 NaN
2023-11-02T07:00:00+00:00           NaN            1.0                 NaN
2023-11-02T10:00:00+00:00           NaN            1.0                 NaN
2023-11-02T11:00:00+00:00           NaN            1.0                 NaN
2023-11-02T14:00:00+00:00           1.0            NaN                 NaN
2023-11-02T15:00:00+00:00           1.0            1.0                 0.0
2023-11-02T15:00:00+00:00           NaN            1.0                 NaN
2023-11-02T17:00:00+00:00           1.0            1.0                 0.0
2023-11-02T18:00:00+00:00           2.0            NaN                 NaN
2023-11-02T20:00:00+00:00           1.0            NaN                 NaN
2023-11-02T21:00:00+00:00           NaN            1.0                 NaN
2023-11-02T23:00:00+00:00           1.0            NaN                 NaN
2023-11-03T01:00:00+00:00           1.0            1.0                 0.0
2023-11-03T05:00:00+00:00           NaN            1.0                 NaN
2023-11-03T07:00:00+00:00           1.0            NaN                 NaN
2023-11-03T08:00:00+00:00           NaN            1.0                 NaN
2023-11-03T08:00:00+00:00           NaN            1.0                 NaN
2023-11-03T09:00:00+00:00           1.0            1.0                 0.0
2023-11-03T12:00:00+00:00           NaN            1.0                 NaN
2023-11-03T13:00:00+00:00           NaN            2.0                 NaN
2023-11-03T14:00:00+00:00           1.0            NaN                 NaN
2023-11-03T15:00:00+00:00           1.0            1.0                 0.0
2023-11-03T16:00:00+00:00           NaN            1.0                 NaN
2023-11-03T17:00:00+00:00           2.0            NaN                 NaN
2023-11-03T19:00:00+00:00           NaN            1.0                 NaN
2023-11-04T02:00:00+00:00           NaN            1.0                 NaN
2023-11-04T09:00:00+00:00           NaN            1.0                 NaN
2023-11-04T09:00:00+00:00           1.0            NaN                 NaN
2023-11-04T10:00:00+00:00           1.0            1.0                 0.0
2023-11-04T14:00:00+00:00           1.0            1.0                 0.0
2023-11-04T15:00:00+00:00           NaN            1.0                 NaN
2023-11-04T19:00:00+00:00           1.0            NaN                 NaN
2023-11-04T21:00:00+00:00           1.0            1.0                 0.0
2023-11-04T23:00:00+00:00           1.0            NaN                 NaN
2023-11-05T00:00:00+00:00           1.0            NaN                 NaN
2023-11-05T08:00:00+00:00           NaN            1.0                 NaN
2023-11-05T11:00:00+00:00           NaN            1.0                 NaN
2023-11-05T13:00:00+00:00           1.0            1.0                 0.0
2023-11-05T13:00:00+00:00           1.0            1.0                 0.0
2023-11-05T14:00:00+00:00           NaN            1.0                 NaN
2023-11-05T23:00:00+00:00           1.0            NaN                 NaN
2023-11-06T04:00:00+00:00           NaN            1.0                 NaN
2023-11-06T06:00:00+00:00           NaN            1.0                 NaN
2023-11-06T08:00:00+00:00           1.0            1.0                 0.0
2023-11-06T11:00:00+00:00           NaN            1.0                 NaN
2023-11-06T12:00:00+00:00           1.0            NaN                 NaN
2023-11-06T14:00:00+00:00           NaN            1.0                 NaN
2023-11-07T07:00:00+00:00           NaN            1.0                 NaN
2023-11-07T09:00:00+00:00           1.0            1.0                 0.0
2023-11-07T11:00:00+00:00           NaN            1.0                 NaN
```

The `NaN` indicates that there is no bucket for that hour. This is occurring because of what I mentioned above, buckets with a count of 0 are omitted from the logs. Also, there are missing adjacent buckets so we cannot verify if there is a reconciliation or not.

**Comment by ayirr7** [comment]
@NisanthanNanthakumar for my other sample above, I did not post the sorted hourly buckets. Here are the results for [this other log event](https://console.cloud.google.com/logs/query;query=jsonPayload.event%3D%22Generics%20Metrics%20Backend%20query%20results%20not%20the%20same%20as%20Errors%20dataset%20query.%22%0Atimestamp%3D%222023-11-03T00:04:48.203401488Z%22%0AinsertId%3D%22tnfdur0wnnkqhdu5%22;cursorTimestamp=2023-11-03T00:04:48.203401488Z;duration=P2D?project=internal-sentry):


```
                    hourBucket  errors_count  metrics_count  (errors - metrics)
2023-10-27T00:00:00+00:00             1            1.0                 0.0
2023-10-27T00:00:00+00:00             1            1.0                 0.0
2023-10-27T15:00:00+00:00             4            4.0                 0.0
2023-10-27T15:00:00+00:00             3            3.0                 0.0
2023-10-27T15:00:00+00:00             3            3.0                 0.0
2023-10-27T15:00:00+00:00            16           16.0                 0.0
2023-10-27T15:00:00+00:00             2            2.0                 0.0
2023-10-27T15:00:00+00:00             2            2.0                 0.0
2023-10-27T16:00:00+00:00             1            1.0                 0.0
2023-10-27T17:00:00+00:00             3            3.0                 0.0
2023-10-27T17:00:00+00:00             3            3.0                 0.0
2023-10-27T18:00:00+00:00             1            NaN                 NaN
2023-10-27T18:00:00+00:00             1            1.0                 0.0
2023-10-27T19:00:00+00:00             1            1.0                 0.0
2023-10-27T19:00:00+00:00             1            1.0                 0.0
2023-10-27T19:00:00+00:00             1            1.0                 0.0
2023-10-27T20:00:00+00:00             2            2.0                 0.0
2023-10-27T20:00:00+00:00             2            2.0                 0.0
2023-10-27T20:00:00+00:00             2            2.0                 0.0
2023-10-27T21:00:00+00:00             1            1.0                 0.0
2023-10-27T21:00:00+00:00            10           10.0                 0.0
2023-10-27T21:00:00+00:00            10           10.0                 0.0
2023-10-27T22:00:00+00:00             2            2.0                 0.0
2023-10-27T22:00:00+00:00             3            3.0                 0.0
2023-10-28T00:00:00+00:00             3            3.0                 0.0
2023-10-28T00:00:00+00:00             2            2.0                 0.0
2023-10-28T00:00:00+00:00            12           12.0                 0.0
2023-10-28T00:00:00+00:00             2            2.0                 0.0
2023-10-30T17:00:00+00:00             9            9.0                 0.0
2023-10-30T17:00:00+00:00             9            8.0                 1.0
2023-10-30T20:00:00+00:00             9            9.0                 0.0
2023-10-30T20:00:00+00:00            10           10.0                 0.0
2023-10-30T20:00:00+00:00            30           30.0                 0.0
2023-10-30T20:00:00+00:00            59           59.0                 0.0
2023-10-30T23:00:00+00:00             1            1.0                 0.0
2023-10-30T23:00:00+00:00             1            1.0                 0.0
2023-10-31T00:00:00+00:00             2            2.0                 0.0
2023-10-31T00:00:00+00:00             2            2.0                 0.0
2023-10-31T01:00:00+00:00             2            2.0                 0.0
2023-10-31T01:00:00+00:00             2            2.0                 0.0
2023-10-31T03:00:00+00:00             1            1.0                 0.0
2023-10-31T03:00:00+00:00             1            1.0                 0.0
2023-10-31T04:00:00+00:00             7            7.0                 0.0
2023-10-31T04:00:00+00:00             7            7.0                 0.0
2023-10-31T05:00:00+00:00             1            1.0                 0.0
2023-10-31T05:00:00+00:00             1            1.0                 0.0
2023-10-31T06:00:00+00:00             4            4.0                 0.0
2023-10-31T06:00:00+00:00             4            4.0                 0.0
2023-10-31T07:00:00+00:00             4            4.0                 0.0
2023-10-31T07:00:00+00:00             4            4.0                 0.0
2023-10-31T14:00:00+00:00           168          168.0                 0.0
2023-10-31T14:00:00+00:00             7            7.0                 0.0
2023-10-31T14:00:00+00:00            33           33.0                 0.0
2023-10-31T14:00:00+00:00            11           11.0                 0.0
2023-10-31T15:00:00+00:00           141          141.0                 0.0
2023-10-31T15:00:00+00:00            39           39.0                 0.0
2023-10-31T15:00:00+00:00             9            9.0                 0.0
2023-10-31T15:00:00+00:00            12           12.0                 0.0
2023-10-31T16:00:00+00:00            13           13.0                 0.0
2023-10-31T16:00:00+00:00             6            4.0                 2.0
2023-10-31T16:00:00+00:00            12           11.0                 1.0
2023-10-31T17:00:00+00:00             1            3.0                -2.0
2023-10-31T17:00:00+00:00             1            1.0                 0.0
2023-10-31T17:00:00+00:00             1            2.0                -1.0
2023-10-31T21:00:00+00:00             1            1.0                 0.0
2023-10-31T21:00:00+00:00             1            1.0                 0.0
2023-10-31T23:00:00+00:00             2            2.0                 0.0
2023-10-31T23:00:00+00:00             4            4.0                 0.0
2023-10-31T23:00:00+00:00             5            5.0                 0.0
2023-11-01T00:00:00+00:00             2            2.0                 0.0
2023-11-01T00:00:00+00:00             2            2.0                 0.0
2023-11-01T06:00:00+00:00            16           16.0                 0.0
2023-11-01T06:00:00+00:00           103          103.0                 0.0
2023-11-01T06:00:00+00:00            16           16.0                 0.0
2023-11-01T06:00:00+00:00            16           16.0                 0.0
2023-11-01T07:00:00+00:00             6            6.0                 0.0
2023-11-01T07:00:00+00:00            38           38.0                 0.0
2023-11-01T07:00:00+00:00             6            6.0                 0.0
2023-11-01T07:00:00+00:00             6            6.0                 0.0
2023-11-01T15:00:00+00:00           137          137.0                 0.0
2023-11-01T15:00:00+00:00             7            7.0                 0.0
2023-11-01T15:00:00+00:00             7            7.0                 0.0
2023-11-01T15:00:00+00:00            21           21.0                 0.0
2023-11-01T17:00:00+00:00             2            2.0                 0.0
2023-11-01T17:00:00+00:00             2            2.0                 0.0
2023-11-01T18:00:00+00:00             1            1.0                 0.0
2023-11-01T18:00:00+00:00             1            1.0                 0.0
2023-11-01T19:00:00+00:00             2            2.0                 0.0
2023-11-01T19:00:00+00:00             5            5.0                 0.0
2023-11-01T20:00:00+00:00            16           16.0                 0.0
2023-11-01T20:00:00+00:00            52           52.0                 0.0
2023-11-01T20:00:00+00:00             9            9.0                 0.0
2023-11-01T20:00:00+00:00            30           30.0                 0.0
2023-11-01T21:00:00+00:00             4            4.0                 0.0
 2023-11-01T21:00:00+00:00             2            2.0                 0.0
2023-11-01T21:00:00+00:00             3            3.0                 0.0
2023-11-01T22:00:00+00:00             3            3.0                 0.0
2023-11-01T22:00:00+00:00             5            5.0                 0.0
2023-11-01T22:00:00+00:00             6            6.0                 0.0
2023-11-02T08:00:00+00:00             4            4.0                 0.0
2023-11-02T08:00:00+00:00             4            4.0                 0.0
2023-11-02T13:00:00+00:00             2            2.0                 0.0
2023-11-02T13:00:00+00:00             2            2.0                 0.0
2023-11-02T13:00:00+00:00             1            1.0                 0.0
2023-11-02T18:00:00+00:00             8            8.0                 0.0
2023-11-02T18:00:00+00:00            36           36.0                 0.0
2023-11-02T18:00:00+00:00            34           34.0                 0.0
2023-11-02T20:00:00+00:00             4            4.0                 0.0
2023-11-02T20:00:00+00:00             5            5.0                 0.0
2023-11-02T20:00:00+00:00             4            4.0                 0.0
2023-11-02T21:00:00+00:00             1            1.0                 0.0
2023-11-02T21:00:00+00:00             1            NaN                 NaN
2023-11-02T21:00:00+00:00             1            1.0                 0.0
2023-11-02T22:00:00+00:00            19           19.0                 0.0
2023-11-02T22:00:00+00:00            24           24.0                 0.0
2023-11-02T22:00:00+00:00            19           19.0                 0.0
2023-11-02T23:00:00+00:00            14           14.0                 0.0
2023-11-02T23:00:00+00:00            14           14.0                 0.0
2023-11-02T23:00:00+00:00            14           14.0                 0.0
```

As we can see, there are also missing adjacent hourly buckets in this sample and there are also some repeated hourly buckets for some reason. 

So in summary from both of these 2 log events so far:
Would it be possible to investigate how we can get adjacent buckets? I am not able to find any example log where there are adjacent buckets.

**Comment by ayirr7** [comment]
So in summary from both of these 2 log events and both of the tabular results so far:
Would it be possible to investigate how we can get adjacent buckets in the logs? I am not able to find any example log event where there are adjacent buckets.

**Comment by ayirr7** [comment]
@NisanthanNanthakumar I don't believe we should be missing any hourly buckets in the logs. We are constantly emitting metrics in the [post_process](https://github.com/getsentry/sentry/blob/master/src/sentry/tasks/post_process.py#L606-L607) task for each error seen in post_process and there are also errors that are constantly being stored.

**Comment by ayirr7** [comment]
@NisanthanNanthakumar I think this is a possibility; though, I think the small differences we are seeing in counts may be expected. I can run the script on more log entries, but so far, what I've seen is that the (errors - metrics) column usually contains very small differences. 

I don't think we can expect a perfect alignment of numbers here, even amongst adjacent buckets. The errors' timestamps are coming from the SDK, which can be very old. For context: Relay validates that each error event's timestamp is maximum 30 days old. This still means that an error event can come into Sentry with a 29 day old timestamp. The timestamp for the metric, however, is determined in post_process itself. I think it is expected that there could be differences in the way the counts are bucketed ultimately. @evanh Curious to hear what your thoughts are as well.

**Comment by ayirr7** [comment]
@NisanthanNanthakumar Thanks for the PR! I added a [comment](https://github.com/getsentry/sentry/pull/59596#issuecomment-1801410919) to your PR, but essentially, I don't think we can expect the adjacent hourly buckets sums to match. 

I explained in the comment as well, but we can take for example the situation that an error event arrives in Relay with a 30 day old timestamp. This is valid. Therefore, assume that the error event has a timestamp of (`now()`- 30 days). By the time this error event reaches post_process, assume the current time is roughly `now()`+ 2 minutes. Then, the metrics payload will contain a timestamp of `now()`+2 minutes. 

As we can see, the errors timestamp and count will be in a much older bucket as compared to the metrics timestamp. They are not in adjacent buckets. So, I don't think we can expect the adjacent sums to match necessarily.

**Comment by ayirr7** [comment]
@NisanthanNanthakumar Yes, I think it makes sense to determine the scale of the issue/size of the discrepancies at this time. I can take a look at your PR tomorrow. Additionally, I was thinking it could be a good idea for some of us to hop onto a call to discuss next steps more thoroughly. Let me know if that would sound good to you!

**Comment by ayirr7** [comment]
It appears that we are pausing work on this for now -- please correct me if otherwise!
